src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2771,13 +2771,26 @@ module Crystal
     end
 
     def visit(node : Rescue)
+      is_exception_type_callback = Proc(Crystal::Type, Bool).new do |type|
+        type.implements?(@program.exception) || type.module?
+      end
+
       if node_types = node.types
         types = node_types.map do |type|
           type.accept self
           instance_type = type.type.instance_type
-          unless instance_type.implements?(@program.exception) || instance_type.module?
+
+          is_exception_type = case instance_type
+                              when UnionType
+                                instance_type.union_types.all? &is_exception_type_callback
+                              else
+                                is_exception_type_callback.call instance_type
+                              end
```

Comment:
**issue:** I think union types should generally be flattened, but I'm not sure if that's always the case. There *might* still be some edge cases where we can encouter nested unions. It shouldn't be hard to implement this generally in a recursive manner where we can be certain to cover all possible cases.
For example:
```cr
private def allowed_type_in_rescue?(type : UnionType)
  type.union_types.all? do |subtype|
    allowed_type_in_rescue?(subtype)
  end
end

private def allowed_type_in_rescue?(type)
  type.implements?(@program.exception) || type.module?
end
```

Reply:
Good call yea, pushed a commit to add some additional specs. Should be good for another review.